src/server/main_service.cc

```diff
@@ -2403,17 +2403,35 @@ VarzValue::Map Service::GetVarzStats() {
   return res;
 }
 
-std::pair<GlobalState, bool> Service::SwitchState(GlobalState from, GlobalState to) {
+GlobalState Service::SwitchState(GlobalState from, GlobalState to) {
   lock_guard lk(mu_);
-  if (global_state_ != from)
-    return {global_state_, false};
+  // Counting switchs to loading state enables us to do several loading processes (replications)
+  // at the same time. Only when all loading processes finish we exit loading state.
+  // When server is in loading state, calling SwitchState to loading will increase
+  // loading_state_counter_. When server is in loading state, calling SwitchState to active will
+  // decrease loading_state_counter_, if loading_state_counter_ is 0 switch to active.
+  if (global_state_ == to && global_state_ == GlobalState::LOADING) {
+    ++loading_state_counter_;
+    return global_state_;
+  }
+  if (global_state_ == from && global_state_ == GlobalState::LOADING &&
+      loading_state_counter_ > 0) {
+    --loading_state_counter_;
+    if (loading_state_counter_ > 0) {
+      return global_state_;
+    }
+  }
+
+  if (global_state_ != from) {
+    return global_state_;
+  }
```

Comment:
for LOADING the from parameter is used as if we are not already in LOADING state we will change the state to LOADING only if global_state_ == from
so if we are in TAKEOVE/SHUTDOWN we will not change to loading because we ask to move from ACTIVE to LOADING

Reply:
After looking more into the code, I think that we dont realy need the from param in the switch state as we will not be able to issue commands if we are not in active mode
I will do the change we RequestLoadingState and will later revisit the SwitchState function in a different PR